src/ring_theory/adjoin/basic.lean

```diff
@@ -102,6 +113,81 @@ lemma adjoint_prod_le (s : set A) (t : set B) :
   adjoin R (set.prod s t) ≤ (adjoin R s).prod (adjoin R t) :=
 adjoin_le $ set.prod_mono subset_adjoin subset_adjoin
 
+lemma adjoin_le_prod (s) (t) :
+  adjoin R (linear_map.inl R A B '' (s ∪ {1}) ∪
+  linear_map.inr R A B '' (t ∪ {1})) ≤ (adjoin R s).prod (adjoin R t) :=
+begin
+  refine adjoin_le (set.union_subset _ _),
+  { rw set.image_union,
+    refine set.union_subset (λ x hx, set_like.mem_coe.2 $ set.mem_prod.2 _)
+      (λ x hx, set_like.mem_coe.2 $ set.mem_prod.2 _),
+    { obtain ⟨y, hy⟩ := hx,
+      rw [←hy.2, set_like.mem_coe, set_like.mem_coe, linear_map.inl_apply],
+      exact ⟨subset_adjoin hy.1, subalgebra.zero_mem _⟩ },
+    { simp only [set.mem_image, set.mem_singleton_iff, linear_map.inl_apply, exists_eq_left] at hx,
+      rw [←hx, set_like.mem_coe, set_like.mem_coe],
+      exact ⟨subalgebra.one_mem _, subalgebra.zero_mem _⟩ } },
+  { rw set.image_union,
+    refine set.union_subset (λ x hx, set_like.mem_coe.2 $ set.mem_prod.2 _)
+      (λ x hx, set_like.mem_coe.2 $ set.mem_prod.2 _),
+    { obtain ⟨y, hy⟩ := hx,
+      rw [←hy.2, set_like.mem_coe, set_like.mem_coe, linear_map.inr_apply],
+      exact ⟨subalgebra.zero_mem _, subset_adjoin hy.1⟩ },
+    { simp only [set.mem_image, set.mem_singleton_iff, exists_eq_left, linear_map.inr_apply] at hx,
+      rw [←hx, set_like.mem_coe, set_like.mem_coe],
+      exact ⟨subalgebra.zero_mem _, subalgebra.one_mem _⟩ } }
+end
+
+lemma adjoint_prod_fst_mem (B) [semiring B] [algebra R B] {s} {x : A} (h : x ∈ adjoin R s) :
+  ((x, 0) : (A × B)) ∈ (adjoin R ((linear_map.inl R A B) '' (s ∪ {1}))) :=
```

Comment:
If you restate this as
```suggestion
 linear_map.inl R A B x ∈ adjoin R (linear_map.inl R A B '' (s ∪ {1})) :=
```
does it make the proof any easier / harder? Is this more general statement true, which means you don't need to prove this separately for `fst` and `snd`?
```suggestion
  f x ∈ (adjoin R (f '' (s ∪ {1}))) :=
```

Reply:
Nice catch! It is true for `f : A →ₗ[R] B` with the additional assumption that `f(a₁ * a₂) = f a₁ * f a₂)`.

Reply:
Now it uses `mem_adjoin_of_map_mul`. Let me see if even `adjoin_le_prod` can be generalized to arbitrary maps.